tests/test_markup/test_parse_inline.py

```diff
@@ -0,0 +1,137 @@
+import os
+
+import docutils
+import pytest
+from docutils import frontend, nodes
+from docutils.io import StringInput
+
+from sphinx import addnodes, parsers
+from sphinx.io import SphinxStandaloneReader
+from sphinx.util.console import strip_colors
+
+docutils_version = pytest.mark.skipif(
+    docutils.__version_info__ < (0, 19), reason="at least docutils 0.19 required"
+)
+
+
+@pytest.fixture()
+def document(app):
+    settings = frontend.get_default_settings(parsers.RSTParser)
+    settings.env = app.builder.env
+    reader = SphinxStandaloneReader()
+    reader.setup(app)
+    reader.source = StringInput(source_path='dummy/document.rst')
+    reader.settings = settings
+    document = reader.new_document()
+    return document
+
+
+@pytest.fixture()
+def parser(app):
+    parser = parsers.RSTParser()
+    parser.set_application(app)
+    return parser
+
+
+@docutils_version()
+@pytest.mark.parametrize(('rst', 'expected'), [
+    (
+        # pep role
+        ':pep:`8`',
+        [addnodes.index, nodes.target, nodes.reference]
+    ),
+    (
+        # rfc role
+        ':rfc:`2324`',
+        [addnodes.index, nodes.target, nodes.reference]
+    ),
+    (
+        # correct interpretation of code with whitespace
+        '``code   sample``',
+        [nodes.literal]
+    ),
+    (
+        # no ampersands in guilabel
+        ':guilabel:`Foo`',
+        [nodes.inline]
+    ),
+    (
+        # kbd role
+        ':kbd:`space`',
+        [nodes.literal]
+    ),
+    (
+        # description list: simple
+        'term\n    description',
+        [nodes.Text]
+    ),
+    (
+        # description list: with classifiers
+        'term : class1 : class2\n    description',
+        [nodes.Text]
+    ),
```

Comment:
@chrisjsewell 
> `Text` also passes definition lists and section titles; its definitely much cleaner to use the proper inline parsing, even if docutils does not make this as easy 😒

After doing some testing, this doesn't seem to be true. I added the above test to check that definition lists are simply parsed as a single Text node and tested with both implementations and the test passes for both.

I just noticed that the current implementation (directly calling `Inliner.parse()`) has some issues with the reporter source being null in the writing phase. I didn't have this issue with the previous implementation (using `initial_state='Text'`), so I think I will come back to it.

Reply:
> I just noticed that the current implementation (directly calling `Inliner.parse()`) has some issues with the reporter source being null in the writing phase.

Turns out I simply didn't get the source the right way.